### PR TITLE
feat: prometheus metrics

### PR DESCRIPTION
**Pull request by jachro**
This PR:
* adds a new `GET /metrics` endpoint  exposing Prometheus metrics to both services
* make the metrics list the standard JVM metrics
* make the metrics list the standard HTTP service metrics (the `search-api` service)
* make the metrics list two custom Redis metrics: the total size of the queues and a number of unprocessed events (the `search-provision` service)

### CODE REVIEW

**Review by eikek — APPROVED** [review]
👍🏼

File: modules/search-api/src/main/scala/io/renku/search/api/HttpApplication.scala [hunk 1]
```diff
@@ -21,28 +21,40 @@ package io.renku.search.api
...
 
-  lazy val router: HttpApp[F] =
+  private lazy val searchAndOperationRoutes =
     Router[F](
       prefix -> (openapi.routes <+> search.routes),
       "/" -> OperationRoutes[F]
-    ).orNotFound
+    )
+
+  lazy val router: HttpApp[F] =
```
[the remark this answers is not in the document]
Comment (jachro) [reply modules/search-api/src/main/scala/io/renku/search/api/HttpApplication.scala:59]: Yes, that's a good point. I'll try to streamline things a bit.

File: modules/search-api/src/main/scala/io/renku/search/api/Microservice.scala [hunk 2]
```diff
@@ -19,17 +19,20 @@
 package io.renku.search.api
 
 import cats.effect.{ExitCode, IO, IOApp}
+import com.comcast.ip4s.port
 import io.renku.logging.LoggingSetup
+import io.renku.search.http.HttpServer
 
 object Microservice extends IOApp:
 
+  private val port = port"8080"
```
[the remark this answers is not in the document]
Comment (jachro) [reply modules/search-api/src/main/scala/io/renku/search/api/Microservice.scala:28]: 👍🏻

File: modules/search-provision/src/main/scala/io/renku/search/provision/HttpApplication.scala [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+object HttpApplication:
+
+  def apply[F[_]: Async: Network](
+      registryBuilder: CollectorRegistryBuilder[F]
+  ): Resource[F, HttpApp[F]] =
+    MetricsRoutes[F](registryBuilder).makeRoutes
+      .map(new HttpApplication[F](_).router)
+
+final class HttpApplication[F[_]: Async](metricsRoutes: HttpRoutes[F])
+    extends Http4sDsl[F]:
```
[the remark this answers is not in the document]
Comment (jachro) [reply modules/search-provision/src/main/scala/io/renku/search/provision/HttpApplication.scala:40]: Yes, we don't

File: modules/http4s-metrics/src/main/scala/io/renku/search/http/metrics/MetricsRoutes.scala [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+final class MetricsRoutes[F[_]: Sync](registryBuilder: CollectorRegistryBuilder[F]):
+
+  def makeRoutes(measuredRoutes: HttpRoutes[F]): Resource[F, HttpRoutes[F]] =
+    for
+      registry <- registryBuilder.makeRegistry
+      metrics <- Prometheus.metricsOps[F](registry, prefix = "server")
+    yield PrometheusExportService(registry).routes <+> Metrics[F](metrics)(measuredRoutes)
+
+  def makeRoutes: Resource[F, HttpRoutes[F]] =
```
[the remark this answers is not in the document]
Comment (jachro) [reply modules/http4s-metrics/src/main/scala/io/renku/search/http/metrics/MetricsRoutes.scala:36]: So the reason why we need `MetricsRoutes` in two flavours is:
* in the case of `search-api` service we collect metrics of the HTTP API (latencies, number of calls, etc) so we pass the `measuredRoutes` (business APIs) for which we'd like to gather the metrics.
* in the case of `search-provision` we don't want any HTTP metrics to be collected but still want to create routes that are exposed by the microservice.
Would that make sense?

[the remark this answers is not in the document]
Comment (eikek) [reply modules/http4s-metrics/src/main/scala/io/renku/search/http/metrics/MetricsRoutes.scala:36]: Ah I see, thank you! Sorry, I somehow couldn't make the connection it refers to the two services

File: modules/http4s-metrics/src/main/scala/io/renku/search/metrics/CollectorRegistryBuilder.scala [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+ */
+
+package io.renku.search.metrics
+
+import cats.effect.{Resource, Sync}
+import cats.syntax.all.*
+import io.prometheus.client.{CollectorRegistry, Collector as JCollector}
+import org.http4s.metrics.prometheus.PrometheusExportService
+
+final class CollectorRegistryBuilder[F[_]: Sync]:
```
[the remark this answers is not in the document]
Comment (jachro) [reply modules/http4s-metrics/src/main/scala/io/renku/search/metrics/CollectorRegistryBuilder.scala]: Right, I tried to make it immutable at first but ended up with code I wasn't happy with. Now, I tried again now and I think it's a bit better.

**Pull request merged**